chapters/VK_KHR_swapchain/wsi.txt

```diff
@@ -466,6 +466,16 @@ ifdef::VK_KHR_swapchain_mutable_format[]
     created with but are supported for at least one of the allowed image
     view formats.
 endif::VK_KHR_swapchain_mutable_format[]
+ifdef::VK_EXT_present_timing[]
+  * ename:VK_SWAPCHAIN_CREATE_ABSOLUTE_TIME_BIT_EXT specifies that when
+    images of the swapchain are presented, a
+    slink:VkAbsolutePresentTimeEXT can: be provided to specify the
+    absolute time they should: be displayed.
+  * ename:VK_SWAPCHAIN_CREATE_RELATIVE_TIME_BIT_EXT specifies that when
+    images of the swapchain are presented, an
+    slink:VkRelativePresentTimeEXT can: be provided to specify the
+    minimum duration they should: be displayed.
+endif::VK_EXT_present_timing[]
```

Comment:
Wrt. to these new swapchain create flags: Maybe i missed it, but will there be corresponding capability bits somewhere, to signal to the application if VK_SWAPCHAIN_CREATE_ABSOLUTE_TIME_BIT_EXT and/or VK_SWAPCHAIN_CREATE_RELATIVE_TIME_BIT_EXT is supported by a given surface to which the new swapchain is going to be attached? So the app could e.g., only request the timestamp feedback bits, but not presentation time scheduling?

Ideally one wants to have both presentation time scheduling support  as provided by these bits, and feedback about actual presentation timing via vkGetPastPresentationTimingEXT(). But some windowing system implementations on some platforms may only provide the infrastructure to support a partial implementation of the VK_EXT_presentation_timing_extension at the moment, at least for a while until windowing system implementations catch up. Even partial enablement of this extension, e.g., only the timing feedback part, would be valuable for some use cases, so it would be nice if apps could query what is supported and make use of that and adapt for missing bits.

E.g., for FRR display modes the Linux DRM/KMS FOSS display stack has all that is needed to support timing and timestamping in fullscreen exclusive direct display mode (leased DRM/KMS outputs on X11 or future Wayland). For VRR mode, the current kernels have the timestamping part in excellent shape, but nothing yet for scheduling of VRR flips at a specific target time. Would be nice to make use of the extension as far as possible on already shipping Linux kernels.

Similar, X11 with DRI3/Present has all that is needed for FRR timing and FRR/VRR timestamping on the FOSS stack, although NVidia's proprietary driver is severely lacking. Nothing for VRR scheduling though. Current X-Server can be put to good use for a partial extension.

On Wayland we have a stable protocol extension named presentation_time since > 5 years, which provides timing feedback on the subset of compositors that implement it (afaik Weston does it well - i was involved in testing/debugging/improvements, wlroots and Sway seem to support it? Some KDE/Kwin fork seems to support it? Googles chromium seems to have support? may imply ChromeOS? / Android?). There isn't any support for scheduling presentation at a specific target time yet. Would be good to be able to at least use the timestamp feedback part on already shipping Wayland compositors.

On macOS 10.15.7 MoltenVK seems to implement the VK_GOOGLE_display_timing extension at least partially on top of some Metal api timing mechanisms.

On Windows-10, i don't know the status, but some DXGKI structs in public docs suggest there may be at least something somewhat usable for timestamp feedback?

Reply:
A few people have requested this. I added a new structure extending VkSurfaceCapabilities2KHR which looks like this:

```
    // Provided by VK_EXT_present_timing
    typedef struct VkPresentTimingSurfaceCapabilitiesEXT {
        VkStructureType    sType;
        void*              pNext;
        VkBool32           presentTimingSupported;
        VkBool32           presentAtAbsoluteTimeSupported;
        VkBool32           presentAtRelativeTimeSupported;
    } VkPresentTimingSurfaceCapabilitiesEXT;
```

I think we might also want to plug into `vkGetPhysicalDeviceSurfacePresentModes2EXT`, but I'm slightly annoyed that it is provided by `VK_EXT_full_screen_exclusive` which I'm reluctant to add as a requirement to this extension, so I will leave this out for now. We need to better specify interaction with each present mode first anyway.

Reply:
I think the  correct way to pull in a bit of functionality from an unrelated EXT is to add it to both extensions. Then you get OR ifdef logic in the spec for that function, and I believe there are existing samples of a way to handle this in the XML as well.